### PR TITLE
Refine reshape_grad and transpose_grad

### PR DESCRIPTION
**Pull request by chengduoZH**
In this PR, I add an output(`XShape`) in reshape_op and transpose_op, which is flagged as `AsIntermediate` and will be used in reshape_grad_op and transpose_grad_op.
`XShape` is only used to carry the shape and lod of `X`, so we don't need to allocate memory for it. In this way, the memory of `X` can be reused when the reshape_op or transpose_op is finished.

I do some experiments in Transformer, after using `XShape`, the peak memory usage reduces 624MB(10041 MB -> 9417 MB)

### CODE REVIEW

**Review by dzhwinter — DISMISSED** [review]
A small question, do we need to change the python code?  
I vaguely remember that python will check the output variable is filled.  Is the intermediate output will be skipped?

**Comment by chengduoZH** [comment]
@dzhwinter 
> I vaguely remember that python will check the output variable is filled. Is the intermediate output will be skipped?

I add a parameter [here](https://github.com/PaddlePaddle/Paddle/pull/13074/files#diff-44128009ae9df7f159ceb5be833c6aa7R36), it can avoid checking the intermediate_out.

File: paddle/fluid/operators/flatten_op.cc [hunk 1]
```diff
@@ -157,6 +157,105 @@ class FlattenGradOp : public framework::OperatorBase {
...
+// FIXME(zcd): flatten2 adds an intermediate output(XShape) based on flatten,
+// the XShape is used to carry the shape and lod of X which will be used in
+// flatten_grad, in this way, the framework can reuse the memory of X
+// immediately the flatten2_op is finished.
+// Considering compatibility issues, we could not fix flatten2_op
+class Flatten2OpInferShape : public FlattenOpInferShape {
+ public:
+  void operator()(framework::InferShapeContext *ctx) const override {
+    FlattenOpInferShape flatten_op_infer_shape;
+    flatten_op_infer_shape(ctx);
```
Comment (reyoung) [inline paddle/fluid/operators/flatten_op.cc]: You means `FlattenOpInferShape::operator()(ctx)`?

Comment (chengduoZH) [reply paddle/fluid/operators/flatten_op.cc]: Yes, I have fixed it.

**Review by reyoung — DISMISSED** [review]

**Review by reyoung — APPROVED** [review]

**Pull request merged**